cardano-api/src/Cardano/Api/Query.hs

```diff
@@ -166,17 +172,51 @@ newtype UTxO era = UTxO (Map TxIn (TxOut era))
   toJSON (UTxO m) = Aeson.Array . Vector.fromList . map convert $ Map.toList m
    where
+    toText :: TxIn -> Text
+    toText txin = case toJSON txin of
+                    Aeson.String txt -> txt
+                    _ -> error "failure"
     convert :: (TxIn, TxOut era) -> Aeson.Value
-    convert ((TxIn txId (TxIx ix), txout)) =
-      let txin = ( Text.decodeUtf8 (serialiseToRawBytesHex txId)
-                 <> "#"
-                 <> Text.pack (show ix)
-                 )
-      in object [ txin .= toJSON txout]
+    convert (txin, txout) = object [ toText txin .= toJSON txout]
+
+
+newtype SerialisedLedgerState era
+  = SerialisedLedgerState (Serialised (Shelley.NewEpochState (ShelleyLedgerEra era)))
+
+data LedgerState era where
+  LedgerState :: ShelleyLedgerEra era ~ ledgerera => Shelley.NewEpochState ledgerera -> LedgerState era
+
+instance (Typeable era, Shelley.TransLedgerState FromCBOR (ShelleyLedgerEra era)) => FromCBOR (LedgerState era) where
+  fromCBOR = LedgerState <$> (fromCBOR :: Decoder s (Shelley.NewEpochState (ShelleyLedgerEra era)))
+
+instance ToJSON (LedgerState ShelleyEra) where
+  toJSON (LedgerState newEpochS) = object [ "lastEpoch" .= Shelley.nesEL newEpochS
+                                          , "blocksBefore" .= Shelley.nesBprev newEpochS
+                                          , "blocksCurrent" .= Shelley.nesBcur newEpochS
+                                          , "stateBefore" .= Shelley.nesEs newEpochS
+                                          , "possibleRewardUpdate" .= Shelley.nesRu newEpochS
+                                          , "stakeDistrib" .= Shelley.nesPd newEpochS
+                                          ]
+
+instance ToJSON (LedgerState AllegraEra) where
+  toJSON (LedgerState newEpochS) = object [ "lastEpoch" .= Shelley.nesEL newEpochS
+                                          , "blocksBefore" .= Shelley.nesBprev newEpochS
+                                          , "blocksCurrent" .= Shelley.nesBcur newEpochS
+                                          , "stateBefore" .= Shelley.nesEs newEpochS
+                                          , "possibleRewardUpdate" .= Shelley.nesRu newEpochS
+                                          , "stakeDistrib" .= Shelley.nesPd newEpochS
+                                          ]
+
 
+instance ToJSON (LedgerState MaryEra) where
+  toJSON (LedgerState newEpochS) = object [ "lastEpoch" .= Shelley.nesEL newEpochS
+                                          , "blocksBefore" .= Shelley.nesBprev newEpochS
+                                          , "blocksCurrent" .= Shelley.nesBcur newEpochS
+                                          , "stateBefore" .= Shelley.nesEs newEpochS
+                                          , "possibleRewardUpdate" .= Shelley.nesRu newEpochS
+                                          , "stakeDistrib" .= Shelley.nesPd newEpochS
+                                          ]
```

Comment:
Can't we make this one instance for any `IsShelleyBasedEra era =>` ?

Reply:
After some fiddling:
```
instance ( IsShelleyBasedEra era
               , ShelleyLedgerEra era ~ ledgerera
               , Consensus.ShelleyBasedEra ledgerera
               , ToJSON (Core.TxOut ledgerera)) => ToJSON (LedgerState era) where
```